### PR TITLE
Truncate OAuth tables after staging DB restore

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Another step towards completing #1989

## Description

This PR adds the truncation steps described in the IP for this work on step 12.3 here: https://docs.openverse.org/projects/proposals/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.html#dag

We also needed a connection to the staging database (we haven't had the need arise for one yet) so I added a new connection for that. The naming convention we have isn't _great_ but I didn't particularly want to change that at this juncture. I've added this connection to our production Airflow instance.

The DAG definition specifies a handful of tables to truncate, but fortunately all of them reference `api_throttledapplication` either directly or indirectly, so a single `TRUNCATE...CASCADE` will truncate them all (and prevents deadlocks which might be incurred by truncating them all in parallel).

image

## Testing Instructions

Testing this is a little troublesome - you'll need to perform all the setup instructions provided in #2099.

Additionally, we can't connect to the _actual_ staging database because we need to connect to the jumphost first, and it's difficult (read: extremely tedious) to try and get the Airflow container to connect to the database which is bound to the host at 5432. The easiest way to test this is to have those commands run on the _local_ API database instead. This can be done by setting `AIRFLOW_CONN_POSTGRES_OPENLEDGER_API_STAGING=postgres://deploy:deploy@db:5432/openledger`. Then run `just api/init` and `just c` to set up the API and start the catalog. You can then kick off the DAG, and the truncation steps should run against the local API database.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by zackkrida** [comment]
> we can't connect to the actual staging database because we need to connect to the jumphost first, and it's difficult (read: extremely tedious) to try and get the Airflow container to connect to the database which is bound to the host at 5432. 

👍 This also seems a little scary, considering how simple it is to connect to the _prod_ DB via the jumphost.

**Review by krysal — COMMENTED** [review]
The changes look good. I may be skipping some steps. Could you copy over here the full instructions to test this? I added the suggested variable but the DAG failed both tasks "get_latest_prod_snapshot" and "get_staging_db_details".

**Review by stacimc — CHANGES_REQUESTED** [review]
I tested this by setting `SKIP_STAGING_DATBASE_RESTORE` to `false`, but did not set up the AWS connection because I was hesitant to run the actual restore and delete operations. This meant that the tasks all failed, so I manually marked them as success and then cleared just the `truncate_tables` task to run it in isolation.

I checked the tables listed in the IP to verify that they had been truncated, and saw that `api_oauth2registration` was not. It looks like this one does not reference `api_throttledapplication` and was not targeted by the CASCADE?

**Comment by AetherUnbound** [comment]
> I tested this by setting `SKIP_STAGING_DATBASE_RESTORE` to `false`, but did not set up the AWS connection because I was hesitant to run the actual restore and delete operations. This meant that the tasks all failed, so I manually marked them as success and then cleared just the `truncate_tables` task to run it in isolation.

Ah nice, @krysal this should be a way to test this piece without the other variables/database restore steps!

> I checked the tables listed in the IP to verify that they had been truncated, and saw that `api_oauth2registration` was not. It looks like this one does not reference `api_throttledapplication` and was not targeted by the CASCADE?

Shoot, I had thought that would be covered by the cascade. I'll double check and if not add that in!

**Comment by AetherUnbound** [comment]
Interesting, it seems we record `api_oauth2registration` and use the `name` in `api_throttledapplication` but not as a foreign key reference 😮 seems like that will also need to be truncated, I'll add that in!

**Review by stacimc — APPROVED** [review]
Looks good!

**Review by krysal — APPROVED** [review]
Tested registering for credentials locally and worked as expected 👏

File: catalog/dags/database/staging_database_restore/staging_database_restore_dag.py [hunk 1]
```diff
@@ -18,13 +18,19 @@
 
 import logging
 from datetime import datetime, timedelta
+from textwrap import dedent as d
```
[the remark this answers is not in the document]
Comment (AetherUnbound) [reply catalog/dags/database/staging_database_restore/staging_database_restore_dag.py:21]: Dedent can be particularly annoying due to how frequently it's often used. I consider it to be similar to the translation aliases that some languages use (I think in Vue or Django it's frequently `_`?). That to say that `d` is an innocuous alias and would be a pattern that's worth trying to follow throughout the project I think.

**Pull request merged**